<div style="border:solid green 2px; padding: 20px">
<b>Hello, Chris!</b>

My name is Grigoriy and I'm going to review your project.

My main goal is not to show that any mistakes have been made somewhere, but to share my experience that will help you in your further work. Further in the file you can see my comments, try to take them into account when performing the next tasks. **Please, don't delete them.**


You can find my comments in <font color='green'>green</font>, <font color='blue'>blue</font> or <font color='orange'>orange</font> boxes like this:

<div class="alert alert-success">
<b>Success:</b> if everything is done succesfully
</div>

<div class="alert alert-warning">
<b>Remarks: </b> if I can give some recommendations
</div>

<div class="alert alert-danger">
<b>Needs fixing:</b> if the block requires some corrections. Work can't be accepted with the red comments.
</div>

    
Let's work on the project in dialogue: if you change something in the project or respond to my comments, write about it. It will be easier for me to track the changes if you highlight your comments:
    
<div class="alert alert-info"> <b>Student comment:</b> For example like this.</div>

### <font color='orange'>General feedback</font>
Thank you for sending your project. You've done a really good job on it!  
I especially liked your work with data functions. Also, there are some comments to help you to improve your project. It doesn't mean that you did something wrong. These comments can help you to further develop your skills.
Your project has passed code-review. Congratulations!


## Analyzing borrowers’ risk of defaulting

Your project is to prepare a report for a bank’s loan division. You’ll need to find out if a customer’s marital status and number of children has an impact on whether they will default on a loan. The bank already has some data on customers’ credit worthiness.

Your report will be considered when building a **credit scoring** of a potential customer. A ** credit scoring ** is used to evaluate the ability of a potential borrower to repay their loan.

<div class="alert alert-warning">
<b>Remarks: </b> It is good practice to add a table of contents to your project. This will allow you not to forget some of its stages and makes a story from your research. <br>
You can find more information here: <a>https://www.notion.so/Project-Presentation-6a8a6e28090b499b9037c48a96ae37c7</a>    
    
</div>

### Step 1. Open the data file and have a look at the general information. 

In [1]:
import pandas as pd #imports the pandas library as function 'pd' so we can process data
from nltk.stem import SnowballStemmer #imports the nltk.stem library so we can process cells according to
#individual words using their stems
english_stemmer = SnowballStemmer('english') #makes the stemmer work with the english language and it will
#be used as 'english_stemmer'
data = pd.read_csv('/datasets/credit_scoring_eng.csv') #reads the the 'credit_scoring_eng.csv' file, makes
#it a DataFrame and saves it to the 'data' variable 
data.info(null_counts=True, verbose=True) #displays the general information of the DataFrame
data #displays the first five and last five rows 

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 21525 entries, 0 to 21524
Data columns (total 12 columns):
children            21525 non-null int64
days_employed       19351 non-null float64
dob_years           21525 non-null int64
education           21525 non-null object
education_id        21525 non-null int64
family_status       21525 non-null object
family_status_id    21525 non-null int64
gender              21525 non-null object
income_type         21525 non-null object
debt                21525 non-null int64
total_income        19351 non-null float64
purpose             21525 non-null object
dtypes: float64(2), int64(5), object(5)
memory usage: 2.0+ MB


children  days_employed  dob_years            education  education_id  \
0             1   -8437.673028         42    bachelor's degree             0   
1             1   -4024.803754         36  secondary education             1   
2             0   -5623.422610         33  Secondary Education             1   
3             3   -4124.747207         32  secondary education             1   
4             0  340266.072047         53  secondary education             1   
...         ...            ...        ...                  ...           ...   
21520         1   -4529.316663         43  secondary education             1   
21521         0  343937.404131         67  secondary education             1   
21522         1   -2113.346888         38  secondary education             1   
21523         3   -3112.481705         38  secondary education             1   
21524         2   -1984.507589         40  secondary education             1   

           family_status  family_status_id gender income_type  debt  \
0                married                 0      F    employee     0   
1                married                 0      F    employee     0   
2                married                 0      M    employee     0   
3                married                 0      M    employee     0   
4      civil partnership                 1      F     retiree     0   
...                  ...               ...    ...         ...   ...   
21520  civil partnership                 1      F    business     0   
21521            married                 0      F     retiree     0   
21522  civil partnership                 1      M    employee     1   
21523            married                 0      M    employee     1   
21524            married                 0      F    employee     0   

       total_income                  purpose  
0         40620.102    purchase of the house  
1         17932.802             car purchase  
2         23341.752    purchase of the house  
3         42820.568  supplementary education  
4         25378.572        to have a wedding  
...             ...                      ...  
21520     35966.698     housing transactions  
21521     24959.969        purchase of a car  
21522     14347.610                 property  
21523     39054.888        buying my own car  
21524     13127.587             to buy a car  

[21525 rows x 12 columns]

<div class="alert alert-warning">
<b>Remarks: </b> Very good that you provide so detailed comments to your code! This habit would help you and your furute co-workers. How you can improve your code:
 <ul>
  <li>It's a good practice to store your import is a separate cell. This is very useful for a code mantaining</li>
  <li>I highly recommend you to format your panads dataframes inputs with head(), tail() or sample(). Different versions of the jupyter notebooks handle with raw dataframes differently. Sometimes this can cause freezes.</li>
</ul>     
</div>

We can hereby see that there are 21525 rows for each of the 12 columns, but some missing values in the 'total_income' and 'days_employed' columns. When we look at the table itself, we also see that the 'days_employed' column has negative values

#### Value Counts for all columns

In [2]:
data['education'].value_counts() #displays all the values the column has and their frequency for the
#education column

secondary education    13750
bachelor's degree       4718
SECONDARY EDUCATION      772
Secondary Education      711
some college             668
BACHELOR'S DEGREE        274
Bachelor's Degree        268
primary education        250
Some College              47
SOME COLLEGE              29
PRIMARY EDUCATION         17
Primary Education         15
graduate degree            4
Graduate Degree            1
GRADUATE DEGREE            1
Name: education, dtype: int64

We have duplicate values in both uppercase and lowercase

In [3]:
data['family_status'].value_counts() #displays all the values the column has and their frequency for the
#'family_status' column

married              12380
civil partnership     4177
unmarried             2813
divorced              1195
widow / widower        960
Name: family_status, dtype: int64

We will also change the name of this column into 'marital_status' and 'family_status_id' into 'marital_status_id'

In [4]:
data['debt'].value_counts() #displays all the values the column has and their frequency for the
#'debt' column

0    19784
1     1741
Name: debt, dtype: int64

The values are 0 (paid on time) and 1 (defaulted)

In [5]:
data['purpose'].value_counts() #displays all the values the column has and their frequency for the
#'purpose' column

wedding ceremony                            797
having a wedding                            777
to have a wedding                           774
real estate transactions                    676
buy commercial real estate                  664
housing transactions                        653
buying property for renting out             653
transactions with commercial real estate    651
purchase of the house                       647
housing                                     647
purchase of the house for my family         641
construction of own property                635
property                                    634
transactions with my real estate            630
building a real estate                      626
buy real estate                             624
purchase of my own house                    620
building a property                         620
housing renovation                          612
buy residential real estate                 607
buying my own car                       

When we look closely at the values in this column, they fall under 4 categories: real estate, car purchase, wedding, and education.

### Conclusion

We need to do something about the missing values in the 'total_income' and 'days_employed' columns. The values in the 'total_income' column are floats. We will get rid of the decimal places. We need to remove the duplicate values in the 'education' column. We need to create a separate table so we can define the meaning of the values in the 'debt' column. We also need to put the purposes into 4 categories: real estate, car purchase, wedding, and education.

<div class="alert alert-success">
<b>Success:</b> Good plan! 
</div>

### Step 2. Data preprocessing

### Processing missing values

Let us take a look at the places in the table where we have missing values

In [6]:
data[data.isna().any(axis=1)]#returns rows which have at least one cell missing a value

children  days_employed  dob_years            education  education_id  \
12            0            NaN         65  secondary education             1   
26            0            NaN         41  secondary education             1   
29            0            NaN         63  secondary education             1   
41            0            NaN         50  secondary education             1   
55            0            NaN         54  secondary education             1   
...         ...            ...        ...                  ...           ...   
21489         2            NaN         47  Secondary Education             1   
21495         1            NaN         50  secondary education             1   
21497         0            NaN         48    BACHELOR'S DEGREE             0   
21502         1            NaN         42  secondary education             1   
21510         2            NaN         28  secondary education             1   

           family_status  family_status_id gender    income_type  debt  \
12     civil partnership                 1      M        retiree     0   
26               married                 0      M  civil servant     0   
29             unmarried                 4      F        retiree     0   
41               married                 0      F  civil servant     0   
55     civil partnership                 1      F        retiree     1   
...                  ...               ...    ...            ...   ...   
21489            married                 0      M       business     0   
21495  civil partnership                 1      F       employee     0   
21497            married                 0      F       business     0   
21502            married                 0      F       employee     0   
21510            married                 0      F       employee     0   

       total_income                   purpose  
12              NaN         to have a wedding  
26              NaN                 education  
29              NaN    building a real estate  
41              NaN  second-hand car purchase  
55              NaN         to have a wedding  
...             ...                       ...  
21489           NaN         purchase of a car  
21495           NaN          wedding ceremony  
21497           NaN       building a property  
21502           NaN    building a real estate  
21510           NaN              car purchase  

[2174 rows x 12 columns]

<div class="alert alert-warning">
<b>Remarks: </b> Too much columns for visual analysis :) I think, <code>data.isna().sum()</code> would be more helpful here.
</div>

In [8]:
data.isna().sum()

children               0
days_employed       2174
dob_years              0
education              0
education_id           0
family_status          0
family_status_id       0
gender                 0
income_type            0
debt                   0
total_income        2174
purpose                0
dtype: int64

Our missing values show up as NaN (Not a Number) which means mathematical operations can be performed on them. They are probably here due to an error when copying the data, or another technical error. We first of all have to decide what to do with the 'days_employed' column. It has both missing values and negative values. Given that this column is not needed for future analysis, it is not worth fixing. So we can drop that column

<div class="alert alert-success">
<b>Success:</b> Very good observation that 'days_employed' has unlogical negative values!
</div>

In [9]:
data.drop(columns = 'days_employed', errors='ignore', inplace=True)
#deletes the 'days_employed' column, ignores all errors, and makes the change on the table without making
#a copy

In [10]:
data.info(null_counts=True, verbose=True)

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 21525 entries, 0 to 21524
Data columns (total 11 columns):
children            21525 non-null int64
dob_years           21525 non-null int64
education           21525 non-null object
education_id        21525 non-null int64
family_status       21525 non-null object
family_status_id    21525 non-null int64
gender              21525 non-null object
income_type         21525 non-null object
debt                21525 non-null int64
total_income        19351 non-null float64
purpose             21525 non-null object
dtypes: float64(1), int64(5), object(5)
memory usage: 1.8+ MB


Looking at the 'total_income' column, there is a little over 2000 missing values. That is a lot. So we won't delete those rows. We will fill them instead. The question is with which value? Two options: median or mean. To know which one to choose, let us take a look at the maximum and minimum values

In [11]:
data['total_income'].max() #greatest value in the 'total_income' column

362496.645

In [12]:
data['total_income'].min() #lowest value in the total_income' column

3306.762

<div class="alert alert-warning">
<b>Remarks: </b> Snall hint: you can get all of this values from one function - describe(). <br>
<code>data['total_income'].describe()</code>        
</div>

With a range of values this great, the right choice will be the median. Before getting the median, it is good to sort the values of the table (in this case, by the 'total_income'). Then we can replace the missing values with the median total income (data['total_income'].median())

<div class="alert alert-success">
<b>Success:</b> You are right, here we need median or mode instead of mean. 
</div>

<div class="alert alert-warning">
<b>Remarks: </b> Please, can you explain your sorting idea? Median values with and without sorting are the same:   
</div>

In [26]:
print('With sort:', data.sort_values(by='total_income', ascending=False).median().loc['total_income'])
print('Without sort:', data.median().loc['total_income'])

With sort: 23202.87
Without sort: 23202.87


In [27]:
data = data.sort_values(by='total_income', ascending=False)
#this sorts the rows of the table in descending order of the 'total_income'
data['total_income'] = data['total_income'].fillna(data['total_income'].median())
#this fills in the missing values (NaN) with the argument of the 'fillna' method, in this case, the median
data.info(null_counts=True, verbose=True) #general info of the table after the changes

<class 'pandas.core.frame.DataFrame'>
Int64Index: 21525 entries, 12412 to 21510
Data columns (total 11 columns):
children            21525 non-null int64
dob_years           21525 non-null int64
education           21525 non-null object
education_id        21525 non-null int64
family_status       21525 non-null object
family_status_id    21525 non-null int64
gender              21525 non-null object
income_type         21525 non-null object
debt                21525 non-null int64
total_income        21525 non-null float64
purpose             21525 non-null object
dtypes: float64(1), int64(5), object(5)
memory usage: 2.0+ MB


No more missing values

### Conclusion

We dropped the 'days_employed' column because it had both missing and wrong values, and it is not needed for future analysis. We replaced the missing values in the 'total_income' column with the median value

<div class="alert alert-success">
<b>Success:</b> 👍
</div>

### Data type replacement

In [29]:
data.info(null_counts=True, verbose=True) #general information about the DataFrame

<class 'pandas.core.frame.DataFrame'>
Int64Index: 21525 entries, 12412 to 21510
Data columns (total 11 columns):
children            21525 non-null int64
dob_years           21525 non-null int64
education           21525 non-null object
education_id        21525 non-null int64
family_status       21525 non-null object
family_status_id    21525 non-null int64
gender              21525 non-null object
income_type         21525 non-null object
debt                21525 non-null int64
total_income        21525 non-null float64
purpose             21525 non-null object
dtypes: float64(1), int64(5), object(5)
memory usage: 2.0+ MB


From the output above, we can see the 'total_income' column contains float values. When we were getting the max and min values earlier, they had 3 decimal places. Those decimal places are not needed so we will drop them by converting the values into integers.

In [30]:
data['total_income']=pd.to_numeric(data['total_income'], errors='raise').astype('int')
#the 'to_numeric' method converts the values in the column in its argument to numerical values; float by default
#the 'astype' method dictates what type of numerical value: 'int' for integer, 'float' for decimals
data['total_income'].head(10) #prints the first 10 cells of the 'total_income' column

12412    362496
19606    352136
9169     276204
20809    274402
17178    273809
17503    255618
18368    248184
18353    228469
15268    216039
11071    205804
Name: total_income, dtype: int64

<div class="alert alert-warning">
<b>Remarks: </b> Good, but what about gender, education_id, family_status, and debt? They could be converted to categoical type, debt - to the boolean.
You can find more information here:<br>
<a>https://pandas.pydata.org/pandas-docs/stable/user_guide/boolean.html</a> <br>
<a>https://pandas.pydata.org/pandas-docs/stable/user_guide/categorical.html</a> 
    
Rule of thumb: try to avoid object data type in your  dataframes.     
</div>

### Conclusion

We converted total income values to integers for simplicity

### Processing duplicates

We remember the uppercase/lowercase duplicates in the 'education' column 

In [35]:
data['education'].value_counts() 

secondary education    13750
bachelor's degree       4718
SECONDARY EDUCATION      772
Secondary Education      711
some college             668
BACHELOR'S DEGREE        274
Bachelor's Degree        268
primary education        250
Some College              47
SOME COLLEGE              29
PRIMARY EDUCATION         17
Primary Education         15
graduate degree            4
Graduate Degree            1
GRADUATE DEGREE            1
Name: education, dtype: int64

In this case, there probably wasn't any standard set on how to put in the values (uppercase or lowercase). The solution to this is to convert all the values to one case, lowercase being the much preferred option. We will use the 'str.lower()' method

In [36]:
data['education'] = data['education'].str.lower()
#applies the 'str.lower()' method to the education column, thereby conerting all its values to lowercase
data['education'].value_counts()

secondary education    15233
bachelor's degree       5260
some college             744
primary education        282
graduate degree            6
Name: education, dtype: int64

<div class="alert alert-success">
<b>Success:</b> Great that you handle string data with str.lower()
</div>

All values in the 'education' column are in lowercase, and therefore there are no duplicates any longer

<div class="alert alert-danger">
<b>Need to fix:</b> Good, but what about other duplicates? We can still find them with <code>data.duplicated().sum()</code>
</div>

In [40]:
data.duplicated().sum()

71

In [ ]:
data = data.drop_duplicates().reset_index(drop=True)

### Conclusion

We have solved the duplicates issue in the 'education' column by converting the values into lowercase using the str.lower() method

### Categorizing Data

We will categorize the data in purpose, total_income, debt, and children columns as they will be important in our analysis

#### Categorizing the 'purpose' column

In [ ]:
data['purpose'].value_counts()

We observed earlier that we can put these purposes into 4 categories: car purchase, wedding, education, and real estate. This is where the 'ntlk.stem' library comes in. It helps stem words. We can pick out key words for each category. For purposes that have to so with car purchases, there are 2 key words: car and cars. We stem them using the 'english_stemmer' function we defined at the top with the stem method taking the word we want to stem as an argument (i.e 'english_stemmer.stem('car')' and 'english_stemmer.stem('cars')'). We can save those values to a list called 'car_stems'. We can do the same for the other categories; we just need to find their keywords. 

Keywords for education: 'education', 'educated', and 'university'
Keywords for real estate: 'estate', 'property', 'house', and 'housing'
Keyword for wedding: 'wedding' (no need to put this one in a list. Saving it to a variable is just fine)

We will stem every word in the in the 'purpose' values and compare them to the stems of the keywords we collected using a conditional statement and return a string value which will be our category name.

<div class="alert alert-success">
<b>Success:</b> Nice introduction! If this task were real, it would be very useful to your readers.
</div>

In [41]:
def purpose_type(purpose):
#this line defines a function we want to create called 'purpose_type' which takes 'purpose' as argument
     
    car_stems = [english_stemmer.stem('car'), english_stemmer.stem('cars')]
#defines a list called 'car_stems' containing the stems of the 'car purchase' keywords
    
    wedding_stem = english_stemmer.stem('wedding')
#defines a variable containing the stem of the word 'wedding' 
    
    education_stems = [english_stemmer.stem('education'), english_stemmer.stem('educated'), english_stemmer.stem('university')]
#defines a list called 'education_stems' containing the stems of all 'education' keywords
    
    real_estate_stems = [english_stemmer.stem('estate'), english_stemmer.stem('property'), english_stemmer.stem('house'), english_stemmer.stem('housing')]
#defines a list called 'real_estate_stems' containing the stems of all 'real estate' keywords
    
    for word in purpose.split(" "):
#this splits the purpose into its individual words separated by a space (" ") and loops through every word (for)
       
        stemmed_word = english_stemmer.stem(word)
#this creates a variable called 'stemmed_word' and saves to it the stem of each word of the purpose as it loops
        
#Conditional statements:

        if stemmed_word in car_stems:
            return 'car purchase'
#If the stemmed word is found in the 'car_stems' list, the string 'car purchase' will be returned

        if stemmed_word == wedding_stem:
            return 'wedding'
#If the stemmed word is equal to the 'wedding_stem' value, the string 'wedding' will be returned

        if stemmed_word in education_stems:
            return 'education'
#If the stemmed word is found in the 'education_stems' list, the string 'education' will be returned

        if stemmed_word in real_estate_stems:
            return 'real estate'
#If the stemmed word is found in the 'real_estate_stems' list, the string 'real estate' will be returned
            
data['purpose'] = data['purpose'].apply(purpose_type)
#Using the 'apply' method with the purpose_type function we just created as an argument
#we apply the function to the purpose column

data['purpose'].value_counts()#displays the purpose column's values and their frequencies 

real estate     10840
car purchase     4315
education        4022
wedding          2348
Name: purpose, dtype: int64

Our categorical values are in place in the purpose column. We could do this replacement without defining a function, but we would have to loop through rows and that would take more processing time and memory. Saving the code to a function and applying it to a single column takes less processing time and less memory

#### Categorizing the 'debt' column

There are 2 values in the debt column: 0 (meaning paid on time) and 1 (meaning defaulted)

In [42]:
data['debt'].value_counts()

0    19784
1     1741
Name: debt, dtype: int64

We need to create a dictionary to explain what those values mean. We first have apply a function (that we will call 'debt_status') to the 'debt' column and save the output to a new column that we will call 'debt_status' 

In [43]:
def debt_status(debt_id):
#this defines a a function called 'debt_status' that will take as argument 'debt_id'

    if debt_id == 1:
        return 'defaulted'
#if the input value is 1, it returns the string value 'defaulted'
    if debt_id == 0:
        return 'on time'
#if the input value is 0, it returns the string value 'on time'
    
data['debt_status'] = data['debt'].apply(debt_status)
#applies the debt_status function to the 'debt' column and saves the results to the new 'debt_status' column

data.head(10) #displays the first 10 rows of the table

children  dob_years            education  education_id  \
12412         0         44    bachelor's degree             0   
19606         1         39    bachelor's degree             0   
9169          1         35  secondary education             1   
20809         0         61  secondary education             1   
17178         0         42    bachelor's degree             0   
17503         0         43  secondary education             1   
18368         1         41    bachelor's degree             0   
18353         1         41    bachelor's degree             0   
15268         1         64    bachelor's degree             0   
11071         1         36    bachelor's degree             0   

           family_status  family_status_id gender income_type  debt  \
12412            married                 0      M    business     0   
19606            married                 0      M    business     1   
9169   civil partnership                 1      M    employee     0   
20809          unmarried                 4      F    employee     0   
17178  civil partnership                 1      M    business     0   
17503            married                 0      M    business     0   
18368  civil partnership                 1      M    business     0   
18353          unmarried                 4      F    business     0   
15268           divorced                 3      M    business     0   
11071  civil partnership                 1      F    employee     0   

       total_income       purpose debt_status  
12412        362496   real estate     on time  
19606        352136   real estate   defaulted  
9169         276204     education     on time  
20809        274402   real estate     on time  
17178        273809       wedding     on time  
17503        255618   real estate     on time  
18368        248184       wedding     on time  
18353        228469  car purchase     on time  
15268        216039   real estate     on time  
11071        205804   real estate     on time

<div class="alert alert-warning">
<b>Remarks: </b> Oh, that's brutal :) Technically you are right, but there are more effective ways to achive your goal.
    
You can convert debt to boolean data type and store values like True and False. After that you can replace your values with numpy.where: <a>https://stackoverflow.com/questions/39698672/pandas-replace-boolean-value-with-string-or-integer</a> <br>

Rule of thumb: try to avoid apply function. It makes your code much slower.
You can read more here: <a>https://realpython.com/fast-flexible-pandas/#pandas-apply</a>
    
<div>


The 'debt_status' column will say 'defaulted' wherever the 'debt' column says '1'; and wherever 'debt' column says '0', 'debt_status' column will say 'on time'. Now we can create the dictionary which is basically a table explaining what a set of values from the main table mean. We will call that dictionary 'debt_status_dict' and we will save into it the values of the 'debt' and 'debt_status' columns from the main 'data' table. After deleting all duplicates, the dictionary is ready

In [44]:
debt_status_dict = data[['debt', 'debt_status']]
#creates a 'debt_status_dict' table and saves to it the 'debt' and 'debt_status' columns from
#the 'data' table

debt_status_dict = debt_status_dict.drop_duplicates().reset_index(drop=True).sort_values(by='debt', ascending=True)
#this drops duplicates from the table, resets the row numbering (index), and sorts it descending order
#of the 'debt' column values

debt_status_dict #displays the 'debt_status_dict' table

debt debt_status
0     0     on time
1     1   defaulted

#### Categorizing 'total_income'

The 'total_income' has a lot of different values

In [45]:
len(data['total_income'].unique()) #gives us the number of unique values in the 'total_income' column

15388

We are definitely NOT creating a dictionary for 19349 categories. These are quantitative values so we can create 2 categories: 1 category for incomes below or equal to a certain value (like the median) and another category for incomes above that same value. We will define a function (we will call it 'income_category'); it will take the 'total_income' as argument and return 'below median' if the income is below or equal to the median, or 'above median' if the income is above the median. We will apply the function to the 'total_income' column. The output will be saved in a new column called 'income_category'

In [46]:
def income_category(total_income):
#creates function called 'income_category' that takes the total_income as argument

    if total_income <= data['total_income'].median():
        return 'below median'
#If total_income is less than or equal to the median total income, return the string 'below median' 
   
    return 'above median'
#If total_income is greater than the median total income, return the string 'above median'

data['income_category'] = data['total_income'].apply(income_category)
#applies our 'income_category' function to the 'total_income' column and saves the output to 
#the new 'income_category' column

data.head(10) #displays the first ten rows of the 'data' table

children  dob_years            education  education_id  \
12412         0         44    bachelor's degree             0   
19606         1         39    bachelor's degree             0   
9169          1         35  secondary education             1   
20809         0         61  secondary education             1   
17178         0         42    bachelor's degree             0   
17503         0         43  secondary education             1   
18368         1         41    bachelor's degree             0   
18353         1         41    bachelor's degree             0   
15268         1         64    bachelor's degree             0   
11071         1         36    bachelor's degree             0   

           family_status  family_status_id gender income_type  debt  \
12412            married                 0      M    business     0   
19606            married                 0      M    business     1   
9169   civil partnership                 1      M    employee     0   
20809          unmarried                 4      F    employee     0   
17178  civil partnership                 1      M    business     0   
17503            married                 0      M    business     0   
18368  civil partnership                 1      M    business     0   
18353          unmarried                 4      F    business     0   
15268           divorced                 3      M    business     0   
11071  civil partnership                 1      F    employee     0   

       total_income       purpose debt_status income_category  
12412        362496   real estate     on time    above median  
19606        352136   real estate   defaulted    above median  
9169         276204     education     on time    above median  
20809        274402   real estate     on time    above median  
17178        273809       wedding     on time    above median  
17503        255618   real estate     on time    above median  
18368        248184       wedding     on time    above median  
18353        228469  car purchase     on time    above median  
15268        216039   real estate     on time    above median  
11071        205804   real estate     on time    above median

<div class="alert alert-warning">
<b>Remarks: </b> I see, you really like apply function :) <br>
Alternative way: <code>data['income_category'] = (data['total_income'] <= data['total_income'].median())</code>                

#### Categorizing the 'children' column

This column tells us how many kids the clients have, so we are dealing with quantitative values again. One of the questions we will have to answer is if having kids has an effect on repaying loans on time. So we have our 2 categories: 'no children' and 'children'

In [48]:
def children_status(children):
#creates function called 'children_category' that takes 'children' as argument
    
    if children == 0:
        return 'no children'
#If children is equal to zero, return the string 'no children' 
    
    if children > 0:
        return 'children'
#If children is greater than zero, return the string 'children'

data['children_status'] = data['children'].apply(children_status)
#applies our 'children_status' function to the 'children' column and saves the output to 
#the new 'children_status' column

data.head(10)#prints the first ten rows of the 'data' table

children  dob_years            education  education_id  \
12412         0         44    bachelor's degree             0   
19606         1         39    bachelor's degree             0   
9169          1         35  secondary education             1   
20809         0         61  secondary education             1   
17178         0         42    bachelor's degree             0   
17503         0         43  secondary education             1   
18368         1         41    bachelor's degree             0   
18353         1         41    bachelor's degree             0   
15268         1         64    bachelor's degree             0   
11071         1         36    bachelor's degree             0   

           family_status  family_status_id gender income_type  debt  \
12412            married                 0      M    business     0   
19606            married                 0      M    business     1   
9169   civil partnership                 1      M    employee     0   
20809          unmarried                 4      F    employee     0   
17178  civil partnership                 1      M    business     0   
17503            married                 0      M    business     0   
18368  civil partnership                 1      M    business     0   
18353          unmarried                 4      F    business     0   
15268           divorced                 3      M    business     0   
11071  civil partnership                 1      F    employee     0   

       total_income       purpose debt_status income_category children_status  
12412        362496   real estate     on time    above median     no children  
19606        352136   real estate   defaulted    above median        children  
9169         276204     education     on time    above median        children  
20809        274402   real estate     on time    above median     no children  
17178        273809       wedding     on time    above median     no children  
17503        255618   real estate     on time    above median     no children  
18368        248184       wedding     on time    above median        children  
18353        228469  car purchase     on time    above median        children  
15268        216039   real estate     on time    above median        children  
11071        205804   real estate     on time    above median        children

#### Categorizing 'family_status_id'

In [49]:
data['family_status_id'].value_counts()

0    12380
1     4177
4     2813
3     1195
2      960
Name: family_status_id, dtype: int64

We would usually need to make another column then export them to the dictionary, but an explanatory column exists already in our table: 'family_status'. Let us get value counts for'family_status'

In [50]:
data['family_status'].value_counts()

married              12380
civil partnership     4177
unmarried             2813
divorced              1195
widow / widower        960
Name: family_status, dtype: int64

We have our values. First of all, let's change "widow / widower" to "widowed". We are doing that for simplicity

In [51]:
data['family_status']=data['family_status'].replace('widow / widower', 'widowed')
#changes the 'family_status' column by replacing every occurrence of 'widow / widower' to 'widowed'
data['family_status'].value_counts()#value occurrences after the change

married              12380
civil partnership     4177
unmarried             2813
divorced              1195
widowed                960
Name: family_status, dtype: int64

Now we can create our dictionary, that we will call 'marital_status_dict', and we will save to it the 'family_status' and 'family_status_id' from the main 'data' table. After removing duplicates, the dictionary is ready

In [52]:
marital_status_dict = data[['family_status', 'family_status_id']]
#creates a table called 'marital_status_dict' and saves to it the 'family_status' and 'family_status_id' columns
#from the main 'data' table

marital_status_dict = marital_status_dict.drop_duplicates().reset_index(drop=True).sort_values(by='family_status_id', ascending=True)
#removes duplicate rows, resets the row numbering (index) and sorts the values in ascending order of the
#'family_status_id' values

marital_status_dict #displays the dictionary

family_status  family_status_id
0            married                 0
1  civil partnership                 1
4            widowed                 2
3           divorced                 3
2          unmarried                 4

### Conclusion

Our 'purpose' values were replaced with categorical 4 values: real estate, car purchase, wedding, and education. Dictionaries were created for the 'debt' and 'family_status_id' with the new 'debt_status' and the already existing 'family_status_id' columns respectively. 'children_status' and 'income_category' were created as category columns for 'children' and 'total_income' columns respectively since they used quantitative values

<div class="alert alert-success">
<b>Success:</b> Good job!
</div>


### Step 3. Answer these questions

- Is there a relation between having kids and repaying a loan on time?

We created a categorical column 'children_status' with 2 values: 'children' and 'no children'. So we will have to group the table by the children_status values and get counts for those who paid on time and those who defaulted. Then we can calculate the default rates for those with kids and those without as percentages. First, the grouping:

In [53]:
children_stats = pd.DataFrame(data.groupby('children_status')['debt_status'].value_counts())
#creates a DataFrame called 'children_stats' where the table 'data' is grouped by 
#the 'children_status' column and counts each occurence of each debt_status value

children_stats #displays the 'children_stats' table

debt_status
children_status debt_status             
children        on time             6652
                defaulted            677
no children     on time            13086
                defaulted           1063

We notice how the 'children_status' and 'debt_status' values are indices. The debt_status counts make the column values. So to locate the amount of customers with children that defaulted, we will use: children_stats.loc['children', 'defaulted']['debt_status']. We will have to divide this by the total number of clients with children i.e children_stats.loc['children'].sum(). The result times 100 is our default rate. A similar process and syntax is used for those with no kids

#### Default rate for those with kids:

In [54]:
default_fraction_kids = children_stats.loc['children', 'defaulted']['debt_status'] / children_stats.loc['children'].sum()
#divides number of customers with children who defaulted by total number of customers with children

default_rate_kids = default_fraction_kids * 100
#calculates percentage

default_rate_kids #displays default rate for customers who have children

debt_status    9.237277
dtype: float64

#### Default rate for those with no kids:

In [55]:
default_fraction_no_kids = children_stats.loc['no children', 'defaulted']['debt_status'] / children_stats.loc['no children'].sum()
#divides number of customers with no children who defaulted by total number of customers with no children

default_rate_no_kids = default_fraction_no_kids*100
#calculates percentage

default_rate_no_kids #displays default rate for customers who have children

debt_status    7.512898
dtype: float64

### Conclusion

According to those findings, customers with kids are more likely to default on their loan (that is, not pay on time) than customers without kids. 

- Is there a relation between marital status and repaying a loan on time?

The categorical column 'family_status' had 5 values: 'married', 'unmarried', 'divorced', 'widowed', and 'civil partnership'. So we will have to group the table by the family_status categories and get counts for those who paid on time and those who defaulted for each category. Then we can calculate the default rates for each category. First, the grouping:

In [56]:
marital_stats = pd.DataFrame(data.groupby('family_status')['debt_status'].value_counts())
#creates a DataFrame called 'marital_stats' where the table 'data' is grouped by 
#the 'family_status' column and counts each occurence of each debt_status value

marital_stats #displays the marital_stats table

debt_status
family_status     debt_status             
civil partnership on time             3789
                  defaulted            388
divorced          on time             1110
                  defaulted             85
married           on time            11449
                  defaulted            931
unmarried         on time             2539
                  defaulted            274
widowed           on time              897
                  defaulted             63

We notice how the 'family_status' and 'debt_status' values are indices. The debt_status counts make the column values. So to locate the amount of customers in a civil partnership that defaulted, we will use: marital_stats.loc['civil partnership', 'defaulted']['debt_status']. We will have to divide this by the total number of clients in a civil partnership i.e children_stats.loc['civil partnership'].sum(). The result times 100 is our default rate. A similar process and syntax is used for the other categories

#### Civil Partnership Default Rate

In [57]:
default_fraction_cp = marital_stats.loc['civil partnership', 'defaulted']['debt_status'] / marital_stats.loc['civil partnership'].sum()
#divides number of customers in a civil partnership who defaulted by total number of customers in
#a civil partnership

default_rate_cp = default_fraction_cp*100
#calculates percentage

default_rate_cp #default rate for those in civil partnership

debt_status    9.288963
dtype: float64

#### Divorced Default Rate

In [58]:
default_fraction_divorced = marital_stats.loc['divorced', 'defaulted']['debt_status'] / marital_stats.loc['divorced'].sum()
#divides number of divorced customers who defaulted by total number of divorced customers

default_rate_divorced = default_fraction_divorced*100
#calculates percentage

default_rate_divorced #displays default rate for divorced customers

debt_status    7.112971
dtype: float64

#### Married Default Rate

In [59]:
default_fraction_married = marital_stats.loc['married', 'defaulted']['debt_status'] / marital_stats.loc['married'].sum()
#divides number of married customers who defaulted by total number of married customers

default_rate_married = default_fraction_married*100
#calculates percentage

default_rate_married #displays default rate for married customers

debt_status    7.520194
dtype: float64

#### Unmarried Default Rate

In [60]:
default_fraction_unmarried = marital_stats.loc['unmarried', 'defaulted']['debt_status'] / marital_stats.loc['unmarried'].sum()
#divides number of unmarried customers who defaulted by total number of unmarried customers

default_rate_unmarried = default_fraction_unmarried*100
#calculates the percentage

default_rate_unmarried #deault rate for unmarried customers

debt_status    9.740491
dtype: float64

#### Widowed Default

In [61]:
default_fraction_widowed = marital_stats.loc['widowed', 'defaulted']['debt_status'] / marital_stats.loc['widowed'].sum()
#divides number of widowed customers who defaulted by total number of widowed customers

default_rate_widowed = default_fraction_widowed*100
#calculates the percentage

default_rate_widowed #displays the default rate for the widowed customers

debt_status    6.5625
dtype: float64

### Conclusion

The highest default rates belong to the "unmarried" and "civil partnership" groups (both over 9%), followed by the married and divorced groups (7.5% and 7.1%). The lowest default rate belongs to the "widowed" group (6.5%)

- Is there a relation between income level and repaying a loan on time?

We created the categorical column 'income_category' with 2 category values: 'above median' and 'below median'. So we will have to group the 'data' table by the 'income_categories' categories and get counts for those who paid on time and those who defaulted for each category. Then we can calculate the default rates for each category. First, the grouping:

In [62]:
income_stats = pd.DataFrame(data.groupby('income_category')['debt_status'].value_counts())
#creates a DataFrame called 'income_stats' where the table 'data' is grouped by 
#the 'income_category' column and counts each occurence of each debt_status value

income_stats #displays the table

debt_status
income_category debt_status             
above median    on time             8908
                defaulted            767
below median    on time            10876
                defaulted            974

We notice how the 'income_category' and 'debt_status' values are indices. The debt_status counts make the column values. So to locate the amount of customers with income above the median that defaulted, we will use: income_stats.loc['above median', 'defaulted']['debt_status']. We will have to divide this by the total number of clients with above-median income i.e income_stats.loc['above median'].sum(). The result times 100 is our default rate. A similar process and syntax is used for the other category

#### Default rate for Above-median income Customers

In [63]:
default_fraction_above_med = income_stats.loc['above median', 'defaulted']['debt_status']/income_stats.loc['above median'].sum()
#divides number of customers with above-median income who defaulted by total number of
#above-median income customers

default_rate_above_med = default_fraction_above_med*100
#calculates the percentage

default_rate_above_med #displays the default rate for customers with above-median income

debt_status    7.927649
dtype: float64

#### Default rate for Below-median income Customers

In [64]:
default_fraction_below_med = income_stats.loc['below median', 'defaulted']['debt_status']/income_stats.loc['below median'].sum()
#divides number of customers with below-median income who defaulted by total number of
#below-median income customers

default_rate_below_med = default_fraction_below_med*100
#calculates the percentage

default_rate_below_med #displays the default rate for customers with above-median income

debt_status    8.219409
dtype: float64

Just to remind ourselves what this median income is

In [65]:
data['total_income'].median()

23202.0

### Conclusion

Customers whose income level is above the median are slightly less likely to default on their loans than those below the median

- How do different loan purposes affect on-time repayment of the loan?

We filled the 'purpose' column with 4 category values: 'wedding', 'car purchase', 'real estate', and 'education'. So we will have to group the 'data' table by the 'purpose' categories and get counts for those who paid on time and those who defaulted for each category. Then we can calculate the default rates for each category. First, the grouping:

In [66]:
purpose_stats = pd.DataFrame(data.groupby('purpose')['debt_status'].value_counts())
#creates a DataFrame called 'purpose_stats' where the table 'data' is grouped by 
#the 'purpose' column and counts each occurence of each debt_status value

purpose_stats #displays the the purpose_stats table

debt_status
purpose      debt_status             
car purchase on time             3912
             defaulted            403
education    on time             3652
             defaulted            370
real estate  on time            10058
             defaulted            782
wedding      on time             2162
             defaulted            186

We notice how the 'purpose' and 'debt_status' values are indices. The debt_status counts make the column values. So to locate the amount of customers whose purpose of taking a loan was car purchase that defaulted, we will use: purpose_stats.loc['car purchase', 'defaulted']['debt_status']. We will have to divide this by the total number of clients who took out loans for car purchase i.e income_stats.loc['car purchase'].sum(). The result times 100 is our default rate. A similar process and syntax is used for the other categories

#### Default Rate for Car Purchase Loans

In [67]:
default_fraction_car_purchase = purpose_stats.loc['car purchase', 'defaulted']['debt_status']/purpose_stats.loc['car purchase'].sum()
#divides number of customers that took out loans for car purchases who defaulted by total number of
#customers who took out loans for car purchases

default_rate_car_purchase = default_fraction_car_purchase*100
#calculates the percentage

default_rate_car_purchase
#displays default rate for car purchase loans

debt_status    9.339513
dtype: float64

#### Default Rate for Education Loans

In [68]:
default_fraction_education = purpose_stats.loc['education', 'defaulted']['debt_status']/purpose_stats.loc['education'].sum()
#divides number of education loans that defaulted by total number of education loans

default_rate_education = default_fraction_education*100
#calculates the percentage

default_rate_education #default rate for education loans

debt_status    9.199403
dtype: float64

#### Default Rate for Real Estate Loans

In [69]:
default_fraction_real_estate = purpose_stats.loc['real estate', 'defaulted']['debt_status']/purpose_stats.loc['real estate'].sum()
#divides number of real estate loans that defaulted by total number of real estate loans

default_rate_real_estate = default_fraction_real_estate*100
#calculates the percentage

default_rate_real_estate #displays the default rate for real estate loans

debt_status    7.214022
dtype: float64

#### Default Rate for Wedding Loans

In [70]:
default_fraction_wedding = purpose_stats.loc['wedding', 'defaulted']['debt_status']/purpose_stats.loc['wedding'].sum()
#divides number of wedding loans that defaulted by total number of wedding loans

default_rate_wedding = default_fraction_wedding*100 #calculates the percentage
default_rate_wedding #displays the default rate for wedding loans

debt_status    7.921635
dtype: float64

### Conclusion

Default rates for car loans and education loans are the highest (above 9%), followed by wedding loans (about 8%), and the lowest being real estate loans (just above 7%)

<div class="alert alert-success">
<b>Success:</b> Excellent work! You were able to highlight key features and provide recommendations. I recommend that you make more focus in your conclusions on the results of your observations than on technical details.
</div>

### Step 4. General conclusion

We have worked with data on customers' income, loan purpose, debt status, family status, and children status. We filled missing values in the total income column with the median, we categorized the loan purposes into car purchase, wedding, education and real estate. We created dictionaries for family status, children, and debt. We grouped the main data table by purpose, family status, children status, and income category. We found the default rates for each category in each of those columns. We found that:

Default rates for car loans and education loans are the highest (above 9%), followed by wedding loans (about 8%), and the lowest being real estate loans (just above 7%); customers whose income level is above the median (i.e 23202) are slightly less likely (7.9%) to default on their loans than those below the median(8.2%); the highest default rates with respect to family status belong to the "unmarried" and "civil partnership" groups (both over 9%), followed by the married and divorced groups (7.5% and 7.1%), the lowest default rate belongs to the "widowed" group (6.5%); and customers customers with kids are more likely(9.2%) to default on their loan (that is, not pay on time) than customers without kids(7.5%). The bank will find the categories with lower default rates more credit worthy i.e No kids, High Income, widowed, and taking a loan for real estate. Least favorable will be those from high default rate categories i.e having kids, low income, unmarried or in a civil partnership, and taking a car or education loan.



<div class="alert alert-success">
<b>Success:</b> God job, Chris! It was a pleasure to review your work.
</div>

### Project Readiness Checklist

Put 'x' in the completed points. Then press Shift + Enter.

- [x]  file open;
- [ ]  file examined;
- [ ]  missing values defined;
- [ ]  missing values are filled;
- [ ]  an explanation of which missing value types were detected;
- [ ]  explanation for the possible causes of missing values;
- [ ]  an explanation of how the blanks are filled;
- [ ]  replaced the real data type with an integer;
- [ ]  an explanation of which method is used to change the data type and why;
- [ ]  duplicates deleted;
- [ ]  an explanation of which method is used to find and remove duplicates;
- [ ]  description of the possible reasons for the appearance of duplicates in the data;
- [ ]  data is categorized;
- [ ]  an explanation of the principle of data categorization;
- [ ]  an answer to the question "Is there a relation between having kids and repaying a loan on time?";
- [ ]  an answer to the question " Is there a relation between marital status and repaying a loan on time?";
- [ ]   an answer to the question " Is there a relation between income level and repaying a loan on time?";
- [ ]  an answer to the question " How do different loan purposes affect on-time repayment of the loan?"
- [ ]  conclusions are present on each stage;
- [ ]  a general conclusion is made.